src/ray/core_worker/store_provider/plasma_store_provider.cc

```diff
@@ -90,14 +99,47 @@ Status CoreWorkerPlasmaStoreProvider::Get(const std::vector<ObjectID> &ids,
     for (size_t i = 0; i < object_buffers.size(); i++) {
       if (object_buffers[i].data != nullptr) {
         const auto &object_id = unready_ids[i];
-        (*results)[unready[object_id]] =
-            std::make_shared<PlasmaBuffer>(object_buffers[i].data);
+        (*results)[unready[object_id]] = std::make_shared<RayObject>(
+            std::make_shared<PlasmaBuffer>(object_buffers[i].data),
+            std::make_shared<PlasmaBuffer>(object_buffers[i].metadata));
         unready.erase(object_id);
+        // TODO (kfstorm): metadata should be structured.
+        std::string metadata = object_buffers[i].metadata->ToString();
+        auto error_type_descriptor = ray::rpc::ErrorType_descriptor();
+        for (int i = 0; i < error_type_descriptor->value_count(); i++) {
+          auto error_type_number = error_type_descriptor->value(i)->number();
+          if (metadata == std::to_string(error_type_number)) {
+            should_break = true;
+          }
+        }
       }
     }
 
     num_attempts += 1;
-    // TODO(zhijunfu): log a message if attempted too many times.
+    if (num_attempts % WARN_PER_NUM_ATTEMPTS == 0) {
+      // Print a warning if we've attempted too many times, but some objects are still
+      // unavailable.
+      std::ostringstream oss;
+      size_t printed = 0;
+      for (auto &entry : unready) {
+        if (printed >= MAX_IDS_TO_PRINT_IN_WARNING) {
+          break;
+        }
+        if (printed > 0) {
+          oss << ", ";
+        }
+        oss << entry.first.Hex();
+      }
+      if (printed < unready.size()) {
+        oss << ", etc";
+      }
+      RAY_LOG(WARNING)
```

Comment:
Will this be printed when waiting for tasks that simply take a very long time to run?

Reply:
Yep, this warning has existed in Java for long time. And we found it very useful when debugging issues.